src/node_file.cc

```diff
@@ -629,130 +634,164 @@ static void Link(const FunctionCallbackInfo<Value>& args) {
   CHECK_NE(*dest, nullptr);
 
   if (args[2]->IsObject()) {  // link(src, dest, req)
-    CHECK_EQ(args.Length(), 3);
+    CHECK_EQ(argc, 3);
     AsyncDestCall(env, args, "link", *dest, dest.length(), UTF8,
                   AfterNoArgs, uv_fs_link, *src, *dest);
-  } else {  // link(src, dest)
-    SYNC_DEST_CALL(link, *src, *dest, *src, *dest)
+  } else {  // link(src, dest, undefined, ctx)
+    CHECK_EQ(argc, 4);
+    fs_req_wrap req_wrap;
+    SyncCall(env, args[3], &req_wrap, "link",
+             uv_fs_link, *src, *dest);
   }
 }
 
 static void ReadLink(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
 
-  const int argc = args.Length();
-
-  CHECK_GE(argc, 1);
+  int argc = args.Length();
+  CHECK_GE(argc, 3);
 
   BufferValue path(env->isolate(), args[0]);
   CHECK_NE(*path, nullptr);
 
   const enum encoding encoding = ParseEncoding(env->isolate(), args[1], UTF8);
 
   if (args[2]->IsObject()) {  // readlink(path, encoding, req)
-    CHECK_EQ(args.Length(), 3);
+    CHECK_EQ(argc, 3);
     AsyncCall(env, args, "readlink", encoding, AfterStringPtr,
               uv_fs_readlink, *path);
-  } else {
-    SYNC_CALL(readlink, *path, *path)
-    const char* link_path = static_cast<const char*>(SYNC_REQ.ptr);
+  } else {  // readlink(path, encoding, undefined, ctx)
+    CHECK_EQ(argc, 4);
+    fs_req_wrap req_wrap;
+    int err = SyncCall(env, args[3], &req_wrap, "readlink",
+                       uv_fs_readlink, *path);
+    if (err) {
+      return;  // syscall failed, no need to continue, error info is in ctx
+    }
+    const char* link_path = static_cast<const char*>(req_wrap.req.ptr);
 
     Local<Value> error;
     MaybeLocal<Value> rc = StringBytes::Encode(env->isolate(),
-                                               link_path,
-                                               encoding,
-                                               &error);
+                                              link_path,
+                                              encoding,
+                                              &error);
```

Comment:
alignment seems off?

Reply:
@targos Yep, I'll fix that later, thanks for catching. Wondering why the cpp linter didn't catch that..